src/Microsoft.Extensions.FileProviders.Physical/PollingWildCardChangeToken.cs

```diff
@@ -0,0 +1,171 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Security.Cryptography;
+using System.Text;
+using Microsoft.Extensions.FileSystemGlobbing;
+using Microsoft.Extensions.FileSystemGlobbing.Abstractions;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.Extensions.FileProviders.Physical
+{
+    /// <summary>
+    /// A polling based <see cref="IChangeToken"/> for wildcard patterns.
+    /// </summary>
+    public class PollingWildCardChangeToken : IChangeToken
+    {
+        // Internal for unit testing.
+        internal static readonly TimeSpan DefaultPollingInterval = TimeSpan.FromSeconds(4);
+        private readonly object _enumerationLock = new object();
+        private readonly DirectoryInfoBase _directoryInfo;
+        private readonly Matcher _matcher;
+        private bool _changed;
+        private DateTime? _lastScanTimeUtc;
+        private byte[] _fileNameBuffer;
+        private byte[] _previousHash;
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="PollingWildCardChangeToken"/>.
+        /// </summary>
+        /// <param name="root">The root of the file system.</param>
+        /// <param name="pattern">The pattern to watch.</param>
+        public PollingWildCardChangeToken(
+            string root,
+            string pattern)
+            : this(
+                new DirectoryInfoWrapper(new DirectoryInfo(root)),
+                pattern,
+                Physical.Clock.Instance)
+        {
+        }
+
+        // Internal for unit testing.
+        internal PollingWildCardChangeToken(
+            DirectoryInfoBase directoryInfo,
+            string pattern,
+            IClock clock)
+        {
+            _directoryInfo = directoryInfo;
+            Clock = clock;
+
+            _matcher = new Matcher(StringComparison.OrdinalIgnoreCase);
```

Comment:
Might result in concurrency issues (since FSW's calls to the `Matcher` might get called in the background while this is running). The type isn't thread safe so it's more work to manage the concurrency.